kvstore/pebble/kvstore.go

```diff
@@ -0,0 +1,163 @@
+	}
+	defer closer.Close()
+	return append([]byte{}, value...), nil
+}
+
+// Delete removes the key-value pair associated with the given key.
+func (store *pebbleKVStore) Delete(key []byte) error {
+	if key == nil {
+		return ErrPebbleUnableToDeleteValue
+	}
+	err := store.db.Delete(key, pebble.Sync)
+	if err != nil {
+		return errors.Join(ErrPebbleUnableToDeleteValue, err)
+	}
+	return nil
+}
+
+// GetAll retrieves all key-value pairs with keys starting with the given prefix.
+// If descending is true, it returns the results in reverse lexicographical order.
+func (store *pebbleKVStore) GetAll(prefix []byte, descending bool) (keys, values [][]byte, err error) {
+	iter, _ := store.db.NewIter(&pebble.IterOptions{
+		LowerBound: prefix,
+		UpperBound: prefixEndBytes(prefix),
+	})
+	defer iter.Close()
+
+	if descending {
+		for valid := iter.Last(); valid; valid = iter.Prev() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	} else {
+		for iter.First(); iter.Valid(); iter.Next() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	}
+
+	if err := iter.Error(); err != nil {
+		return nil, nil, errors.Join(ErrPebbleIteratingStore, err)
+	}
+
+	return keys, values, nil
+}
+
+// Exists checks if a key exists in the store and has a non-empty value.
+func (store *pebbleKVStore) Exists(key []byte) (bool, error) {
+	value, closer, err := store.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, errors.Join(ErrPebbleUnableToGetValue, err)
+	}
+	defer closer.Close()
+	return len(value) > 0, nil
+}
+
+// ClearAll removes all key-value pairs from the store.
+func (store *pebbleKVStore) ClearAll() error {
+	iter, _ := store.db.NewIter(nil)
+	defer iter.Close()
+	for iter.First(); iter.Valid(); iter.Next() {
+		if err := store.db.Delete(iter.Key(), pebble.Sync); err != nil {
+			return errors.Join(ErrPebbleClearingStore, err)
+		}
+	}
+	if err := iter.Error(); err != nil {
+		return errors.Join(ErrPebbleClearingStore, err)
+	}
+	return nil
+}
+
+// Stop closes the database connection.
+func (store *pebbleKVStore) Stop() error {
+	return store.db.Close()
+}
+
+// Len returns the number of key-value pairs in the store.
+func (store *pebbleKVStore) Len() (int, error) {
+	count := 0
+	iter, _ := store.db.NewIter(nil)
+	defer iter.Close()
+	for iter.First(); iter.Valid(); iter.Next() {
+		count++
+	}
+	if err := iter.Error(); err != nil {
+		return 0, errors.Join(ErrPebbleGettingStoreLength, err)
+	}
+	return count, nil
+}
+
+// prefixEndBytes returns the end key for prefix scans.
+func prefixEndBytes(prefix []byte) []byte {
+	if len(prefix) == 0 {
+		return nil
+	}
+	end := make([]byte, len(prefix))
+	copy(end, prefix)
+	for i := len(end) - 1; i >= 0; i-- {
+		end[i]++
+		if end[i] != 0 {
+			return end[:i+1]
+		}
+	}
```

Comment:
Could you add a comment explaining why incrementing the end bytes is done or the purpose in general

Reply:
For the same reason [we do this in the badger store](https://github.com/pokt-network/smt/blob/main/kvstore/badger/kvstore.go#L182) - so we can catch all the keys with that prefix.
Here's an [example in pebble code](https://github.com/cockroachdb/pebble/blob/c2749cc7e49fa691b6727f8d55810aec5bc777e6/iterator_example_test.go#L65).

Reply:
@okdas Do you mind #PUC with:
1. Those links and potentially
2. A symbolic example of the bytes (in hex) before and after?

For 2, I'd personally use a unit test to visualize it in case that helps.

Reply:
@Olshansk added a comment with an example and a test.